src/utilities/extendSchema.js

```diff
@@ -241,30 +237,29 @@ export function extendSchema(
     const existingDirectives = schema.getDirectives();
     invariant(existingDirectives, 'schema must have default directives');
 
-    const newDirectives = directiveDefinitions.map(directiveNode =>
-      definitionBuilder.buildDirective(directiveNode),
+    return existingDirectives.concat(
+      directiveDefinitions.map(node => astBuilder.buildDirective(node)),
     );
-    return existingDirectives.concat(newDirectives);
-  }
-
-  function getTypeFromDef<T: GraphQLNamedType>(typeDef: T): T {
-    const type = definitionBuilder.buildType(typeDef.name);
-    return (type: any);
   }
 
-  // Given a type's introspection result, construct the correct
-  // GraphQLType instance.
-  function extendType(type: GraphQLNamedType): GraphQLNamedType {
-    if (isObjectType(type)) {
-      return extendObjectType(type);
-    }
-    if (isInterfaceType(type)) {
-      return extendInterfaceType(type);
-    }
-    if (isUnionType(type)) {
-      return extendUnionType(type);
+  function extendType<T: GraphQLNamedType>(type: T): T {
+    let extendedType = extendTypeCache[type.name];
+
+    if (!extendedType) {
+      if (isIntrospectionType(type)) {
+        extendedType = type;
+      } else if (isObjectType(type)) {
+        extendedType = extendObjectType(type);
+      } else if (isInterfaceType(type)) {
+        extendedType = extendInterfaceType(type);
+      } else if (isUnionType(type)) {
+        extendedType = extendUnionType(type);
+      } else {
+        extendedType = type;
+      }
+      extendTypeCache[type.name] = extendedType;
     }
-    return type;
+    return (extendedType: any);
```

Comment:
I suggest factoring this into two functions, `getExtendedType` and `extendType`, the first one performing caching (with a comment that extendType need only be called once per type). That should allow you to remove or at least further isolate the `any` cast.

Reply:
@leebyron Great idea 👍 I will make the change in a few minutes.

Reply:
Cast through any internally while keeping the correct function definition, and we'll report it to the flow team

Reply:
@leebyron Done. Can you please review?